doc/ACL_pytest.md

```diff
@@ -0,0 +1,326 @@
+        loganalyzer/
+```
+
+- The *acl* directory here created, so that other test modules related to ACL test can be put here and ACL related fixtures, etc. can live in conftest.py
+- The *common* has common functionallity required by a set of tests.
+- *acstests* simply a symbolic link to acs PTF test scripts
+
+
+#### test_acl.py
+
+The current test is divided into 4 cases:
+- Basic test - setup ACL rules, run traffic, teardown ACL rules
+- Incremental test - setup ACL rules incrementaly, run traffic, teardown ACL rules
+- Reboot test - setup ACL rules, reboot, run traffic, teardown ACL rules
+- Port toggle test - setup ACL rules, toggle all ports, run traffic, teardown ACL rules
+
+The test logic will be seperated into generic traffic test in *BaseAclTest* and derivatives are meant to provide fixture to setup/teardown according to their requirements.
+
+According to above cases 4 test classes has to be coded:
+
+- *TestAcl*
+- *TestIncrementalAcl*
+- *TestAclWithReboot*
+- *TestAclWithPortToggle*
+
+
+In common for all above cases we need to do the following setup/teardown:
+- setup test variables according to topology and minigraph facts
+e.g. get the list of ToR/Spine ports, setup names and configuration for ACL tables and so on
+- generate and deploy configuration files to the DUT, after the test remove configuration and cleanup
+
+Also configuration generation and dependent traffic run tests will be parametrized for 'ingress', 'egress' ACL tests
+
+##### Test fixtures
+
+As above described setup and configuration generation can be done in scope of *test_acl.py* module only once.
+
+These fixtures, by the way,can live in *acl/conftest.py*, so other ACL test modules can reuse them.
+
+###### Module scope
+
+```python
+@pytest.fixture(scope='module')
+def setup(dut, testbed):
+    '''
+    Setup fixture gathers all test required information from DUT facts and testbed
+    '''
+
+@pytest.fixture(scope='module', params=['ingress', 'egress'])
+def stage(request):
+    '''
+    Small fixture to return the ACL table stage is beeing tested
+    '''
+
+@pytest.yield_fixture(scope='module')
+def config(request, dut, setup):
+    '''
+    Fixture generates ACL tables, ACL rules configuration and deploys on the DUT
+    The return value is config files generated
+    '''
+
+@pytest.yield_fixture(scope='module')
+def table(dut, config, stage):
+    '''
+    Create an ACL table on DUT
+    '''
+
+```
+
+Based on the above it is assumed that tests in one module will reuse the ACL table created by 'table' fixture.
+
+If, for example, some one is developing ACL table scale test he will need a seperate test module ('test_acl_scale.py') and override 'config' fixutre to generate many tables configuration and reuse 'table' fixture.
+
+###### Class scope
+
+
+```python
+
+class BaseAclTest(object):
+    __meta__ = abc.ABCMeta
+
+    @abc.abstractmethod
+    def setup_rules(self):
+        pass
+
+    @abc.abstractmethod
+    def teardown_rules(self):
```

Comment:
The teardown_rules and rules can be combined into one fixture. Just use the `yield` statement in the fixture. Code after the `yield` statement will be executed after test is executed and will be served for teardown purpose.

Reply:
Yes, however right now I think teardown_rules is the same for all derivatives. We can implement teardown_rules in base class. So that's why I seperated it here.